File: backend/app/models/message.py
import uuid

from sqlalchemy import (
    Column,
    DateTime,
    ForeignKey,
    String,
    Text,
)
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from app.db.base import Base


class Message(Base):
    __tablename__ = "messages"

    id = Column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
        index=True,
    )

    conversation_id = Column(
        UUID(as_uuid=True),
        ForeignKey("conversations.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    role = Column(
        String(20),
        nullable=False,
        comment="user | assistant | system | tool",
    )

    content = Column(Text, nullable=True)

    # Flexible bag for token counts, model name, tool calls, LangGraph step info, etc.
    metadata_ = Column(
        "metadata",
        JSONB,
        nullable=True,
        server_default="{}",
    )

    created_at = Column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )

    conversation = relationship(
        "Conversation",
        back_populates="messages",
    )

File: admin/lib/admin/schema/message.ex
defmodule Admin.Schema.Message do
  use Ecto.Schema

  @primary_key {:id, Ecto.UUID, autogenerate: false}
  @foreign_key_type Ecto.UUID

  schema "messages" do
    field :conversation_id, Ecto.UUID
    field :role, :string
    field :content, :string
    field :metadata, :map

    field :created_at, :utc_datetime
  end
end
